addons/missionTest/functions/fnc_onMissionSave.sqf

```diff
@@ -0,0 +1,141 @@
+private _floatingUnits = [];
+{
+    if (((getPos _x) select 2) > 0.5) then {
+        WARNING_3("[%1:%2] is floating %3", _x, typeOf _x, getPos _x);
+        _floatingUnits pushBack _x;
+    };
+} forEach _allUnits;
+if (!(_floatingUnits isEqualTo [])) then {
+    _problems pushBack ["Possible Fall Damage (Soldier placed above ground level)", _floatingUnits];
+};
+
+
+// Verify loadouts exist for all units:
+private _classesFallback = [];
+private _classesNone= [];
+{
+    private _faction = toLower faction _x;
+    private _unitClassname = [typeOf _x] call EFUNC(assignGear,cleanPrefix);
+    private _path = missionConfigFile >> "CfgLoadouts" >> _faction >> _unitClassname;
+
+    if ((!isClass(_path)) && EGVAR(assignGear,useFallback)) then {
+        _path = missionConfigFile >> "CfgLoadouts" >> _faction >> "fallback";
+        if (isClass _path) then {
+            WARNING_1("[%1] is using fallback", typeOf _x);
+            _classesFallback pushBackUnique (typeOf _x);
+        } else {
+            WARNING_1("[%1] is has no fallback", typeOf _x);
+            _classesNone pushBackUnique (typeOf _x);
+        };
+    };
+    if (!isClass(_path)) exitWith {};
+} forEach _allUnits;
+if (!(_classesFallback isEqualTo [])) then {
+    _problems pushBack ["Minor: Unit placed that are using fallback", _classesFallback];
+};
+if (!(_classesNone isEqualTo [])) then {
+    _problems pushBack ["Unit placed that have no loadout", _classesNone];
+};
+
+
+// Verify units have weapons:
+private _checkWeapons = [];
+private _checkMagazines = [];
+{
+    private _unit = _x;
+    if ((side _unit) != civilian) then {
+        if ((primaryWeapon _unit) == "") then {
+            if ((typeOf _unit) == "potato_msv_pol") exitWith {};
+            WARNING_1("[%1] has no rifle", typeOf _unit);
+            _checkWeapons pushBackUnique (typeOf _unit);
+        };
+        {
+            private _weapon = _x;
+            if (_x != "") then {
+                private _ammoCount = 0;
+                private _weaponMags = (getArray (configFile >> "cfgWeapons" >> _weapon >> "magazines")) apply {toLower _x};
+                {
+                    _x params ["_xMagName", "_xAmmo"];
+                    if ((toLower _xMagName) in _weaponMags) then {
+                        _ammoCount = _ammoCount + _xAmmo;
+                    };
+                } forEach (magazinesAmmoFull _unit);
+                if (_ammoCount == 0) then {
+                    if ((toLower _weapon) in ["rhs_weap_rpg26", "rhs_weap_m136", "launch_nlaw_f"]) then {
+                        // TRACE_1("ignoring special AT reloads",_weapon);
+                    } else {
+                        WARNING_2("[%1]'s %2 has zero ammo", typeOf _unit, _weapon);
+                        _checkMagazines pushBackUnique (typeOf _unit);
+                    };
+                };
+                if (_weapon == (primaryWeapon _unit)) then {
+                    private _desiredAmmo = switch (true) do {
```

Comment:
Since this is in a loop we can probably optimize it to a call

```SQF
private _desiredAmmo = [toLower typeOf _unit] call {
    params ["_unitType"];

    // allow low ammo count for long range gunners
    if (_unitType find "marksman" > 0 
            || {_unitType find "spotter" > 0}
            || {_unitType find "sniper" > 0}) exitWith {
        20
    };

    // suggest hight ammo count for MGs
    if (_unitType find "_ar" > 0
            || _unitType find "_mg" > 0
            || _unitType find "_mmgg" > 0) exitWith {
        250
    };

    // default rifleman case
    100
};
```